File: java/midp/com/nokia/mid/ui/frameanimator/FrameAnimator.java
package com.nokia.mid.ui.frameanimator;

public class FrameAnimator {

    public static final int FRAME_ANIMATOR_MAX_CONCURRENT = 5;
    public static final int FRAME_ANIMATOR_VERTICAL = 0;
    public static final int FRAME_ANIMATOR_HORIZONTAL = 1;
    public static final int FRAME_ANIMATOR_FREE_ANGLE = 2;
    public static final int FRAME_ANIMATOR_FRICTION_LOW = 0;
    public static final int FRAME_ANIMATOR_FRICTION_MEDIUM = 1;
    public static final int FRAME_ANIMATOR_FRICTION_HIGH = 2;

    protected int actionType;
    protected int actionID;

    protected static int _numRegistered;

    protected FrameAnimatorListener _registered;

    public FrameAnimator() {
        _registered = null;
    }

    public boolean register(int x, int y, short maxFps, short maxPps, FrameAnimatorListener listener) {
        _registered = listener;
        _numRegistered++;
        return true;
    }

    public void unregister() {
        _registered = null;
        _numRegistered--;
    }

    public void drag(int x, int y) {
        _registered.animate(this, x, y, (short)0, (short)0, (short)0, true);
    }

    public void kineticScroll(int speed, int direction, int friction, float angle) {
        _registered.animate(this, 0, 0, (short)0, (short)0, (short)0, true);
    }

    public void limitedKineticScroll(int speed, int direction, int friction, float angle, int limitUp, int limitDown) {
        _registered.animate(this, 0, 0, (short)0, (short)0, (short)0, true);
    }

    public void stop() {
        // Do nothing; all animations in this implementation are currently synchronous
    }

    public boolean isRegistered() {
        return _registered != null;
    }

    public static int getNumRegisteredFrameAnimators() {
        return _numRegistered;
    }

}

File: midp/frameanimator.js
/* -*- Mode: Java; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
/* vim: set shiftwidth=2 tabstop=2 autoindent cindent expandtab: */

'use strict';

var FrameAnimator = function() {};

FrameAnimator.numRegistered = 0;

FrameAnimator.prototype._isRegistered = false;

FrameAnimator.prototype.register = function(x, y, maxFps, maxPps, listener) {
  this.x = x;
  this.y = y;
  this.maxFps = maxFps;
  this.maxPps = maxPps;
  this.listener = listener;
  this._isRegistered = true;
  ++FrameAnimator.numRegistered;
};

FrameAnimator.prototype.unregister = function() {
  this.x = null;
  this.y = null;
  this.maxFps = null;
  this.maxPps = null;
  this.listener = null;
  this._isRegistered = false;
  --FrameAnimator.numRegistered;
};

FrameAnimator.prototype.isRegistered = function() {
  return this._isRegistered;
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.init.()V"] = function(ctx, stack) {
  var _this = stack.pop();
  _this.nativeObject = new FrameAnimator();
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.register.(IISSLcom/nokia/mid/ui/frameanimator/FrameAnimatorListener;)Z"] = function(ctx, stack) {
  var listener = stack.pop(), maxPps = stack.pop(), maxFps = stack.pop(), y = stack.pop(), x = stack.pop(), _this = stack.pop();

  if (_this.nativeObject.isRegistered()) {
    ctx.raiseExceptionAndYield("java/lang/IllegalStateException", "FrameAnimator already registered");
  }

  if (!listener) {
    ctx.raiseExceptionAndYield("java/lang/NullPointerException", "listener is null");
  }

  if (x < -65535 || x > 65535 || y < -65535 || y > 65535) {
    ctx.raiseExceptionAndYield("java/lang/IllegalArgumentException", "coordinate out of bounds");
  }

  // XXX return false if FrameAnimator.numRegistered >= FRAME_ANIMATOR_MAX_CONCURRENT

  _this.nativeObject.register(x, y, maxFps, maxPps, listener);
  stack.push(1);
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.unregister.()V"] = function(ctx, stack) {
  var _this = stack.pop();

  if (!_this.nativeObject.isRegistered()) {
    ctx.raiseExceptionAndYield("java/lang/IllegalStateException", "FrameAnimator not registered");
  }

  _this.nativeObject.unregister();
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.drag.(II)V"] = function(ctx, stack) {
  var y = stack.pop(), x = stack.pop(), _this = stack.pop();
  console.warn("FrameAnimator.drag(II)V not implemented (" + x + ", " + y + ")");
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.kineticScroll.(IIIF)V"] = function(ctx, stack) {
  var angle = stack.pop(), friction = stack.pop(), direction = stack.pop(), speed = stack.pop(), _this = stack.pop();
  console.warn("FrameAnimator.kineticScroll(IIIF)V not implemented (" +
               speed + ", " + direction + ", " + friction + ", " + angle + ")");
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.limitedKineticScroll.(IIIFII)V"] = function(ctx, stack) {
  var limitDown = stack.pop(), limitUp = stack.pop(), angle = stack.pop(), friction = stack.pop(),
      direction = stack.pop(), speed = stack.pop(), _this = stack.pop();
  console.warn("FrameAnimator.limitedKineticScroll(IIIFII)V not implemented (" +
               speed + ", " + direction + ", " + friction + ", " + angle + ", " + limitUp + ", " + limitDown + ")");
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.stop.()V"] = function(ctx, stack) {
  var _this = stack.pop();
  console.warn("FrameAnimator.stop()V not implemented");
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.isRegistered.()Z"] = function(ctx, stack) {
  var _this = stack.pop();
  stack.push(_this.nativeObject.isRegistered() ? 1 : 0);
};

Native["com/nokia/mid/ui/frameanimator/FrameAnimator.getNumRegisteredFrameAnimators.()I"] = function(ctx, stack) {
  stack.push(FrameAnimator.numRegistered);
};
